packages/rich-text-structure/src/create.js

```diff
@@ -0,0 +1,285 @@
+ * @param {Function}    $3.removeNodeMatch      Function to declare whether the
+ *                                              given node should be removed.
+ * @param {Function}    $3.unwrapNodeMatch      Function to declare whether the
+ *                                              given node should be unwrapped.
+ * @param {Function}    $3.filterString         Function to filter the given
+ *                                              string.
+ * @param {Function}    $3.removeAttributeMatch Wether to remove an attribute
+ *                                              based on the name.
+ *
+ * @return {Object} A rich text record.
+ */
+function createRecord( element, range, $3 = {} ) {
+	if ( ! element ) {
+		return {
+			value: {
+				formats: [],
+				text: '',
+			},
+			selection: {},
+		};
+	}
+
+	const {
+		removeNodeMatch = () => false,
+		unwrapNodeMatch = () => false,
+		filterString = ( string ) => string,
+		removeAttributeMatch,
+	} = $3;
+
+	const filterStringComplete = ( string ) => filterString( string.replace( '\n', '' ) );
+
+	if (
+		element.nodeName === 'BR' &&
+		! removeNodeMatch( element ) &&
+		! unwrapNodeMatch( element )
+	) {
+		return {
+			value: {
+				formats: [ undefined ],
+				text: '\n',
+			},
+			selection: {},
+		};
+	}
+
+	if ( ! element.hasChildNodes() ) {
+		return {
+			value: {
+				formats: [],
+				text: '',
+			},
+			selection: {},
+		};
+	}
+
+	return Array.from( element.childNodes ).reduce( ( accumulator, node ) => {
+		const { formats } = accumulator.value;
+
+		if ( node.nodeType === TEXT_NODE ) {
+			if ( range ) {
+				if ( node === range.startContainer ) {
+					accumulator.selection.start = accumulator.value.text.length + filterStringComplete( node.nodeValue.slice( 0, range.startOffset ) ).length;
+				}
+
+				if ( node === range.endContainer ) {
+					accumulator.selection.end = accumulator.value.text.length + filterStringComplete( node.nodeValue.slice( 0, range.endOffset ) ).length;
+				}
+			}
+
+			const text = filterStringComplete( node.nodeValue, accumulator.selection );
+			accumulator.value.text += text;
+			formats.push( ...Array( text.length ) );
+		} else if ( node.nodeType === ELEMENT_NODE ) {
+			if ( removeNodeMatch( node ) ) {
+				return accumulator;
+			}
+
+			if ( range ) {
+				if (
+					node.parentNode === range.startContainer &&
+					node === range.startContainer.childNodes[ range.startOffset ]
+				) {
+					accumulator.selection.start = accumulator.value.text.length;
+				}
+
+				if (
+					node.parentNode === range.endContainer &&
+					node === range.endContainer.childNodes[ range.endOffset ]
+				) {
+					accumulator.selection.end = accumulator.value.text.length;
+				}
+			}
+
+			let format;
+
+			if ( ! unwrapNodeMatch( node ) && node.nodeName !== 'BR' ) {
+				const type = node.nodeName.toLowerCase();
+				const attributes = getAttributes( node, { removeAttributeMatch } );
+
+				format = attributes ? { type, attributes } : { type };
```

Comment:
I have a spilt off branch that reuses these objects to reduce the size of the structure (so it takes less memory and performs better). I'll add it to this branch ASAP. Another way to do this would be referencing them with an ID and having a list of used formats by ID, then it would also be a smaller object as JSON, but there's no use case for storing this structure elsewhere atm... This is similar btw to how blocks are stored in the global state where there's a list of blocks by ID and a list of the block order. Cc @gziolo @mcsf re performance.

Reply:
Actually, I already did this, but added a test to check in 35bf8b8.